site24x7/oauth/oauth.go

```diff
@@ -0,0 +1,187 @@
+	baseAccurl := fmt.Sprintf("%s%s", tokenDomain, tokenApi)
+	urlToReturn := fmt.Sprintf("%s?%s", baseAccurl, urlValues.Encode())
+	return urlToReturn
+}
+
+type Authenticator struct {
+	// Protects tkns.AccessToken (updated in refresh goroutine)
+	sync.Mutex
+	tkns      tokens
+	storePath string
+}
+
+func (ator *Authenticator) getClient() *http.Client {
+	trTls11 := &http.Transport{
+		DisableKeepAlives: true,
+		TLSClientConfig: &tls.Config{
+			MaxVersion:         tls.VersionTLS13,
+			InsecureSkipVerify: true,
+		},
+	}
+	client := http.Client{
+		Transport: trTls11,
+		Timeout:   timeout,
+	}
+	return &client
+}
+
+func (ator *Authenticator) setDefaultHeaders(request *http.Request) {
+	request.Header.Set("Accept", "application/x-www-form-urlencoded")
+}
+
+func (ator *Authenticator) getAccessTokenFrom(urlToPost string) error {
+	var result map[string]interface{}
+	req, err := http.NewRequest("POST", urlToPost, nil)
+	if err != nil {
+		return err
+	}
+	setRequestHeaders(req, ator.tkns.AccessToken)
+	client := ator.getClient()
+
+	res, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	err = json.NewDecoder(res.Body).Decode(&result)
+	if err != nil {
+		return err
+	}
+	ator.Lock()
+	ator.tkns.AccessToken = result["access_token"].(string)
+	ator.Unlock()
+
+	ator.tkns.ExpiresInSec = result["expires_in_sec"].(float64)
+	if ator.tkns.AccessToken == "" || ator.tkns.ExpiresInSec == 0 {
+		return errors.New("error while fetching access token from refresh token: empty token or no expiration")
+	}
+	ator.tkns.TokenGenerationTime = time.Now().UnixNano() / 1e6
+	return ator.tkns.persist(ator.storePath)
+}
+
+func (ator *Authenticator) getAccessTokenFromRefreshToken() error {
+	return ator.getAccessTokenFrom(getUrl(ator.tkns.refreshTokenURLValues()))
+}
+
+func (ator *Authenticator) setAccessTokenFromCode() error {
+	return ator.getAccessTokenFrom(getUrl(ator.tkns.generatedCodeURLValues()))
+}
+
+func fileExists(path string) (bool, error) {
+	fi, err := os.Lstat(path)
+	if err == nil {
+		if fi.IsDir() {
+			return false, errors.New(path + " is a directory")
+		}
+		return true, nil
+	}
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+	return false, err
+}
+
+func (ator *Authenticator) refresh() error {
+	if ator.tkns.AccessToken == "" {
+		return errors.New("no access token to refresh")
+	}
+	err := ator.getAccessTokenFromRefreshToken()
+	if err != nil {
+		// if we failed to refresh we want to try again in 5 min
+		ator.tkns.ExpiresInSec = 300
+		return err
+	}
+	return nil
+}
+
+func (ator *Authenticator) scheduleRefresh() {
+	go func() {
+		for {
+			timer := time.NewTimer(time.Second * time.Duration(math.Max(ator.tkns.ExpiresInSec - 10, 2)))
+			<-timer.C
+			err := ator.refresh()
+			if err != nil {
+				fmt.Printf("failed to refresh access token: %b", err)
+			}
+		}
+	}()
+}
+
+// AccessToken returns the access token.
+func (ator *Authenticator) AccessToken() string {
+	ator.Lock()
```

Comment:
What about using [`sync.RWMutex`](https://pkg.go.dev/sync?tab=doc#RWMutex)?

Reply:
it doesn't really matter :-)

Reply:
done